A first version of having the possibility to create mobile apps

tsteur opened the pull request:
refs #7893 
fixes #7803 

New wording 
- Type: Can be a "Website", "Mobile App", later a car,  a server, ...
- Properties: So far we only had "Websites". We will use the term "Properties" as soon as multiple types are enabled and / or if multiple types are actually used. It depends where it is used. In the future we will no longer work based on `Sites` but on `Properties`. Not for now though...

This is only a first version since we will need this feature for a Piwik Pro plugin soon. To make it really nice we need to refactor the Settings API (which let's you currently create PluginSettings only, so everything is related to plugins, not properties). Also we need to change many other things and this is not doable in a short time frame. Therefore I made it more or less "work". It will be a bit nicer for Piwik 3.0.0 when we can also change APIs and spend more time on it. None of the new `PropertySettings` API etc will be actually public API as we will remove/change it very soon again.

There are new plugins `WebsiteType` and `MobileAppType`. Let me now re naming. I can name it only `Website` and `MobileApp` but we might want to use this for other things in the future. I do not really mind. `MobileAppType` is disabled by default so far most users nothing really changes. I do not think we will announce this MobileAppType thing big is it does not yet have a real value. This will be added over time.

Once multiple types are enabled we use "Add new property" (in Sites Management only so far) and we will let a user choose what type to create. The PropertySettings API let's you define custom settings for properties. Eg the type MobileAppType defines a custom setting "App ID". On top plugins can add custom settings for all properties or only settings for a certain type.

@mattab you might want to have a look re the descriptions in the Types. I will leave a note in the code.

![create_property](https://cloud.githubusercontent.com/assets/273120/7999963/2db9856c-0ba8-11e5-85e5-5418332be0f4.png)
![edit_mobileapp](https://cloud.githubusercontent.com/assets/273120/7999964/2dba355c-0ba8-11e5-8da5-64923a92f41b.png)

tsteur committed:
refs #7893 added possibility to measure mobile apps

tsteur commented on core/Db/Schema/Mysql.php:
```diff
@@ -102,6 +102,14 @@ public function getTablesCreateSql()
                             ) ENGINE=$engine DEFAULT CHARSET=utf8
             ",
 
+            'site_setting'    => "CREATE TABLE {$prefixTables}site_setting (
```
We will store property settings in this table. The `setting_value` will be serialized as a `Setting` can define a type and we do not want to lose that type, eg an array.

mnapoli replied:
Will it be serialized using `serialize()`? Unless it's really not practical it would be good to find alternatives instead of storing serialized strings: it is only readable by PHP, the format could change (e.g. even today a Piwik install cannot be migrated from PHP to HHVM because of that: not same serialization format), etc.

tsteur replied:
It will be serialized for now yes. We might change it later but since it can store anything and we do need to keep the type of the `setting_value` for now. `json` etc would lose the type under circumstances eg `Array` vs `Objects`.

It doesn't store any instances there. Only `array`, `stdClass`, `int`, `float`, ...

tsteur commented on core/Plugin/Manager.php:
```diff
@@ -1334,6 +1336,10 @@ private function sortPluginsSameOrderAsGlobalConfig(array $plugins)
         }
+
+        // we need to make sure a possibly disabled plugin will be still loaded before any 3rd party plugin
+        $global = array_merge($global, $this->corePluginsDisabledByDefault);
```
I had to add this as I created a 3rd party plugin that required the plugin `MobileAppType`. The 3rd party plugin was always enabled before `MobileAppType` so the missing dependency check did not quite work and the 3rd party plugin was always automatically disabled.

tsteur commented on core/Property/Property.php:
```diff
@@ -0,0 +1,32 @@
+class Property extends Site
+{
+
+    public function getSettingValue($name)
```
I currently do not plan to make this method available in `Site` but can do if needed later.

tsteur commented on core/Property/PropertySetting.php:
```diff
@@ -0,0 +1,70 @@
+
+        $this->writableByCurrentUser = Piwik::isUserHasSomeAdminAccess();
+        $this->readableByCurrentUser = Piwik::isUserHasSomeViewAccess();
+    }
```
At this time we do not yet have a `idSite` which is kinda a trade off so here we only check for general permissions. Before saving we actually check whether a user has admin access for a selected site. In general we always check for permissions when reading the value anyway.

tsteur commented on core/Property/PropertySettings.php:
```diff
@@ -0,0 +1,101 @@
+    protected function init()
+    {
+        $type = Type::getType($this->idType);
+        $type->configurePropertySettings($this);
```
A type, such as `MobileApp` can specify property settings. Eg `App ID`

tsteur commented on core/Property/PropertySettings.php:
```diff
@@ -0,0 +1,101 @@
+         * @param \Piwik\Type\Type $type
+         * @param int $idSite
+         */
+        Piwik::postEvent('Property.initPropertySettings', array($this, $type, $this->idSite));
```
Plugins will be able to specify settings too. For example we will need this in a plugin for Apple App Store username & password setting but also for #8021

tsteur commented on core/Property/Settings/Storage.php:
```diff
@@ -0,0 +1,104 @@
+        $sql   = "DELETE FROM $table WHERE `idsite` = ?";
+        $bind  = array($this->idSite);
+
+        $this->db->query($sql, $bind);
```
Didn't create a separate class for the SQL stuff as we will refactor it soon anyway.

tsteur commented on plugins/CoreHome/angularjs/common/directives/dialog.js:
```diff
@@ -29,7 +29,9 @@
                 element.on( "dialogclose", function() {
-                    scope.$apply($parse(attrs.piwikDialog).assign(scope, false));
+                    setTimeout(function () {
+                        scope.$apply($parse(attrs.piwikDialog).assign(scope, false));
```
Prevent possible JS error

tsteur commented on plugins/MobileAppType/lang/en.json:
```diff
@@ -0,0 +1,7 @@
+  "MobileAppType": {
+    "MobileApp": "Mobile App",
+    "MobileApps": "Mobile Apps",
+    "MobileAppDescription": "A native mobile app for eg iOS, Android, .."
```
@mattab you might want to have a look here re translation

mattab replied:
Suggested description: `A native mobile app for iOS, Android or Titanium. Measure user interactions with one of our SDKs.` ?

tsteur replied:
I'd rather use something like `A native mobile app for iOS, Android or any other mobile operating system`. `Titanium Mobile` is a framework but not an operating system. There are many other mobile platforms out there so it's not only limited to iOS & Android. Just because we only have SDKs for them doesn't mean they are not tracking something else :) Also we wouldn't have to update the description when there are more SDKs

tsteur commented on plugins/Morpheus/templates/settingsMacros.twig:
```diff
@@ -0,0 +1,124 @@
+{% macro singleSetting(setting, index = 0) %}
```
We're now reusing the code of pluginSettings in propertySettings

tsteur commented on plugins/SitesManager/Controller.php:
```diff
@@ -33,8 +35,29 @@ public function index()
-    public function getGlobalSettings() {
+    public function getCustomTypeSettings()
+    {
+        $idSite = Common::getRequestVar('idSite', 0, 'int');
```
We render this here server side with a little delay for simplicity. Otherwise we'll have to create an API that returns information for all settings and we need to render pluginSettings and propertySettings client side etc. We will change this soon and do everything client side

tsteur commented on plugins/SitesManager/angularjs/sites-manager/sites-manager-site.controller.js:
```diff
@@ -77,6 +87,16 @@
                 }, 'GET');
             }
 
+            var settings = $('.typeSettings fieldset').serializeArray();
```
for now we use jQuery to get all the fields, we will change this later when we render the settings client side

tsteur commented on plugins/SitesManager/templates/sites-list/site-fields.html:
```diff
@@ -78,6 +71,11 @@
             <input type="text" ng-model="site.name"/>
         </div>
 
+        <div class="form-group typeSettings"
```
Here we render all property settings server side

tsteur commented on plugins/WebsiteType/lang/en.json:
```diff
@@ -0,0 +1,7 @@
+  "WebsiteType": {
+    "Website": "Website",
+    "Websites": "Websites",
+    "WebsiteDescription": "A website or a mobile web app that is usually consumed via a browser"
```
@mattab you might want to have a look here re translation. Also further below in `plugin.json`

mattab replied:
Maybe `A website consists of web pages typically served from a single web domain.` ?

mnapoli commented:
I haven't had a look in details yet:

- I don't get the name "property", I don't see what it is a property of? Could we use something like "application"? (e.g. track a web application, mobile application, …) or even "project" even though it's generic. But I feel very lost by "property".

> There are new plugins WebsiteType and MobileAppType. Let me now re naming. I can name it only Website and MobileApp but we might want to use this for other things in the future.

+1 for naming them with the `Type` suffix as that's what they are. A `Website` object should represent a website, not the "website type". I'd like Piwik to have objects instead of arrays for representing model "structures", so `Website` would be such a class.

tsteur commented:
We thought about many possible names for that and none of them was really good. Application is not really good as it could be anything, even a Retail Store. In the end we went with `Properties` so far as GA uses this naming as well so it will be at least familiar for many users and we reckon they had a lot of thoughts about the best naming for this. 

Something generic is actually good since it is a name for something generic, although not sure if "Projects" is better. Maybe there are some other suggestions by anyone?

mattab commented on plugins/MobileAppType/plugin.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "MobileAppType",
+  "description": "Allows you to create Mobile Apps"
```
suggested: `Analytics for Mobile: lets you create Mobile Apps properties in Piwik, and measure user interactions with a SDK for iOS, Android or Titanium.`

tsteur replied:
What about something like `Analytics for Mobile: lets you manage 'Mobile App' properties and creates an optimized perspective of your mobile data.`. The SDKs are rather irrelevant and what the type actually does is presenting the data in a different perspective by showing different reports, using different naming, etc.

mattab commented on plugins/WebsiteType/plugin.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "WebsiteType",
+  "description": "Allows you to create websites"
```
suggested: `Analytics for the web: lets you create 'Website' properties.`

diosmosis commented:
Haven't looked much, quick wording suggestions:

* Instead of `Type`, how about `Trackable`?
* Instead of `Properties`, how about `TypeProperties` (or `TrackableProperties`)? If `Properties` are meant to describe properties of `Type`s, then adding that context to the word will make it clearer what it is.

Also, I cannot tell what the difference between Properties & Types are. Is one a word and the other a class name?

Can you provide in the ticket description an overview of the entire system?

mnapoli commented:
@diosmosis If I understand correctly a `Type` is the same as a class, and a `Property` is the instance of the Type (i.e. same as object).

The thing is I don't see what they are a property of? When I lookup the [definition](https://en.wiktionary.org/wiki/property) of the word, the closest thing is:

> An attribute or abstract quality associated with an individual, object or concept

Or also:

> An editable or read-only parameter associated with an application, component or class, or the value of such a parameter.

I find this word very confusing as it suggests there is a larger entity to which properties are attached (and properties "describe" this larger entity).

How about `Resource`? This is a term used in [GA's documentation](https://support.google.com/analytics/answer/2649554?hl=en).

> Google Analytics understands a property only as a resource associated with your tracking code. **When you track a resource using Google Analytics**, you include a property ID in the tracking code that you put on your web pages or in your app source code.

mattab commented:
>>  An attribute or abstract quality associated with an individual, object or concept
> The thing is I don't see what they are a property of? 
eg. a website property is an abstract quality of not an individual (or organisation or business)
 
> How about Resource? This is a term used in GA's documentation.

Resource is not bad, we didn't consider this one before. is it used consistently across GA doc instead of Property, ie. do they use both terms for same meaning?

tsteur commented:
> Instead of Type, how about Trackable?

Not really. In the future we will not only have Trackable things. We will also import data etc. 

> Can you provide in the ticket description an overview of the entire system?

What do you mean of the entire system? I haven't added "much" yet so not sure what do describe.

> How about Resource?

That could work. I think they see "Property" rather as "Possessions/Goods/Wares/Things/Movables/..." it's also a synonym for Resource. As a developer we might see this rather from a technical view which makes it more confusing for us than for users. The closest definition that you mentioned are only the closest from a technical view, not from an everyday English equivalent.

When choosing a name please also keep in mind that we do not only want to have tracked "things" but also imported "things" etc.

> @diosmosis If I understand correctly a Type is the same as a class, and a Property is the instance of the Type (i.e. same as object). E.g. type could be "Mobile App", and property could be "Uber" (which is a mobile app)

Yep that's pretty much it. A property can be associated with any type, eg "Mobile App".

mattab commented:
btw: It may be that Properties make more sense in the GA world because they have the notion of Account and Property becomes "part" of an account. Whereas in Piwik we don't have the notion of Account so maybe Property makes less sense for us.

From [GA doc:](https://support.google.com/analytics/answer/1009618?hl=en#Properties)

> A property is a website, mobile application, or device (e.g. a kiosk or point-of-sale device.) An account can contain one or more properties.
> Within an Analytics account, you add the properties from which you want to collect data.

tsteur commented:
Are you saying `Account` would make more sense? Wouldn't that be kinda confusing with user accounts etc.? I see it rather that they had to use some wording like `Account` to group them. 

BTW: We might have "accounts" in the future as well. Why? Imagine we track Websites, Servers, ... at some point. This could be an interesting way to connect "Properties". Eg you might find a slow avg. response time for a particular page, then a click on the page shows you more details about the servers that are listed under the same account etc. We might solve this differently though, but that's maybe too far away to think about it now

mattab commented:
> Are you saying Account would make more sense?

No, I was saying that Properties may make more sense in the context of Google Analytics because they have the notion of Account. Because at PIwik we don't have notion of Account, `Property` is maybe less obvious a good name as it is for GA. That said I still can't think of better name (`Resource` is not bad but don't think it's better?!). 

[naming is hard](http://hilton.org.uk/presentations/naming)

tsteur commented:
The definition of `Resource` also does not seem really matching. I don't have a strong opinion but If in doubt I would maybe go with the definition that people might already know.

mnapoli commented:
Both of you pointed to different definitions for "property":

- abstract quality of an individual/company/…
- possession/goods

I am very skeptical regarding this name, it doesn't speak anything to me and seems hard to define what it means in the context of Piwik (judging from your answers). I also asked Clemence to get an idea about this and she had no idea what this word would represent, in english or french (even though she works with GA a lot). I'm afraid it would only make sense for hardcore users of GA (analytics professionals), and it would render Piwik more complex for the vast majority of users. (actually a simple google search about GA properties yield blog articles like this: http://chrisburbridge.com/google-analytics-understanding-accounts-vs-properties-vs-profiles/ -> the conclusion + comments are critical about the interface and wording)

Here are other suggestions:

- Sources or Data sources ([example 1](http://www.bimeanalytics.com/?lang=en#datasources), [example 2](http://www.gooddata.com/bi-software/platform), [example 3](http://www.easy-insight.com/data.html)), which make sense e.g. if you also consider a "log import" data source
- Systems

We could also take more natural approach: instead of having a "Properties" menu item (whatever the word will be), that leads to a "Properties" title + "Properties list" and a "Add a new property" button, we could name the feature e.g. "Piwik Connect" (or "Integration", the word isn't so important). I've seen that pattern several times while looking at other analytics SaaS today. That allows to make it more natural to users (focus on the actual content rather that pushing artificial concepts on them). To illustrate I've done a quick mockup (just so that you see what I mean, it is by no means complete or realistic):

![capture d ecran 2015-06-09 a 13 29 26](https://cloud.githubusercontent.com/assets/720328/8056873/cc2e47da-0eac-11e5-9ef8-a346bd53a58c.png)

![capture d ecran 2015-06-09 a 13 33 12](https://cloud.githubusercontent.com/assets/720328/8056872/cc2a4252-0eac-11e5-90fb-e29ea6e6373d.png)

Anyway yes we are getting out of the scope of this single change here, but this change here affects the naming and concepts core to Piwik, and what will happen next, so I think it's better to discuss this now. Maybe it would also be better to involve other people from Piwik Pro (e.g. maybe the marketing team). And yes with this solution we still have to find a name, it's just that it's less important for users as we can talk of the "Piwik Connect" feature rather than the "Property/Data Source/…" concept.

Regarding `Type`, I believe we can also go for a more explicit name. Here are some ideas:

- Monitors
- [Collectors](http://help.logicmonitor.com/developers-guide/add-collector/)

tsteur commented:
> I also asked Clemence to get an idea about this and she had no idea what this word would represent, in english or french (even though she works with GA a lot).

One should come along this term in the admin of GA. It's one of three main actions. As said in the beginning the name is not ideal but haven't seen something that is much better yet.

>  (actually a simple google search about GA properties yield blog articles like this: http://chrisburbridge.com/google-analytics-understanding-accounts-vs-properties-vs-profiles/ -> the conclusion + comments are critical about the interface and wording)

GA is confusing there. But that's because they have Accounts and Profiles which we don't have.

I don't find `Data sources` so good and it might be also used in a different context eg I associate it rather for when selecting a data source (3rd party system like a Shop, Salesforce, ...) to import data.

`Systems` is in my mind somehow something completely different and won't match eg a "Retail store" or an "App" etc.

`Connect` could maybe work although it's rather a "verb". How would we use that in the code base? I mean instead of `new Site($idsite=1)`? Would we do `new Connection($connectionId=1)`? It's also pretty abstract and could be associated with different things.

The synonyms of `possessions` & `property` describe it best for me so far: 
From http://synonyme.woxikon.de/synonyme-englisch/possessions.php & http://synonyme.woxikon.de/synonyme-englisch/property.php
> possessions, property, belongings, goods, assets, things, objects

One major goal we have is to also track wearables, the Internet of things. "Thing" is pretty much what we want. It would be not such a good naming though. I think as we're looking for a very generic term that can represent anything no matter what name we're looking for will be very clear since we're looking for something abstract / generic like "Thing" that on top should work for persons, companies, organizations etc. I can understand why GA picked that name. I'm sure it would be very much less confusing if it wasn't a term in programming as well. In the end we're looking for a better word for "Thing" but can't name it like that. As mentioned I don't have a strong opinion on the naming but find none really good so far which is because of the problem mentioned that we're looking for something very abstract/generic that can be anything. It will be always confusing unless we change something there. 

> Regarding Type, I believe we can also go for a more explicit name. Here are some ideas: Monitors,  Collectors

I don't think those words describe it actually and I would never associate it with this. They are not really any of this. A type is more like a different "View" or perspective of your data. It will enable/disable reports, it will use different wordings, it might cause more or less data to be tracked etc. In the end we have a "Thing" that is of a certain "Type". Very generic as well I know. I picked the wording "Type" for now as it is used in the `piwik_site` table already. Eg if we use the wording "Property", it could be later a "PropertyType" as suggested by @diosmosis . Those classes are not public API and won't be public API before Piwik 3.0.0 so we don't have to decide it yet.

> Maybe it would also be better to involve other people from Piwik Pro (e.g. maybe the marketing team). 

Good point. When asking it will be just very important to also be aware of the vision we want to achieve here etc. We want to provide a universal / generic analytics platform that can track anything from Retail Stores to Websites (Servers, Cars, Apps, Smart devices, ...)

tsteur commented:
I'm wondering how we make progress here? @mattab next step would be get some feedback from Piwik PRO?

diosmosis commented:
> One should come along this term in the admin of GA. It's one of three main actions. As said in the beginning the name is not ideal but haven't seen something that is much better yet.

Resource sounds better, IMO. It's a more neutral word for something someone owns than Property, which connotes both ownership AND value. In other words, saying you can track 'properties' sounds like you can only track 'things you paid for' or 'things that make you money', neither of which has anything to do w/ tracking things (as I assume the Piwik platform will in the future define it). Using 'Property' creates cognitive dissonance for me (even after the clarification that it relates to a possession).

> When asking it will be just very important to also be aware of the vision we want to achieve here etc. We want to provide a universal / generic analytics platform that can track anything from Retail Stores to Websites (Servers, Cars, Apps, Smart devices, ...)

I know it's been stated you don't like it, and I don't really have a vested interest in what wording is used, but I think 'Trackable' is still the best option. It's not a real noun, but it essentially means 'thing that can be tracked', which is exactly what you're looking for, no?

tsteur commented:
> I know it's been stated you don't like it, and I don't really have a vested interest in what wording is used, but I think 'Trackable' is still the best option. It's not a real noun, but it essentially means 'thing that can be tracked', which is exactly what you're looking for, no?

I kinda like `Trackable`. Even better than `Resource`, but just find it might be "wrong" / "not as matching word" for things that are not tracked but imported from eg Salesforce, Shops like Shopify, Magento, Retail Stores, ... 

The question might be whether we would list such "things" (trackable things and imported things) under the same management etc. Most likely we would (I reckon) as one would just enter a Login/Password under that "Property/Resource" and reports would be generated based on imported data. For a user it's not really important whether it's tracked or imported I reckon.

> Resource sounds better, IMO. It's a more neutral word for something someone owns than Property, which connotes both ownership AND value. 

`Resource` is the second best name for me so far and would be okay with that. As said I don't have a strong opinion on it in general. I didn't find the description on Wikipedia 100% matching but none of our wordings will be 100% matching I reckon (apart from "Things" haha). You're the native English speaker so I reckon it's good to trust you on that :)

From http://en.wikipedia.org/wiki/Resource
> A resource is a source or supply from which benefit is produced. Typically resources are materials, energy, services, staff, knowledge, or other assets that are transformed to produce benefit and in the process may be consumed or made unavailable.

diosmosis commented:
>  imported from eg Salesforce, Shops like Shopify, Magento, Retail Stores, ... 

What data would be imported? My only frame of reference is web server logs, but this data is still tracked visits, just in a different format. What else would be imported?

> I didn't find the description on Wikipedia 100% matching but none of our wordings will be 100% matching I reckon (apart from "Things" haha). You're the native English speaker so I reckon it's good to trust you on that :)

The way my intuitive understanding of the words goes, is that a resource isn't necessarily something you pay for, however a property seems like something someone paid for at some point. Furthermore, a property isn't necessarily something you could could track (eg, a stuffed animal) or something you want to track (eg, a brony), whereas tracking resources sounds like a useful thing to do, regardless of what it is.

I could go on, but it's hard to communicate the nuances of a language when so much of it depends on the culture surrounding the language, hopefully my comments give some context.

Another side note about 'Trackable', it might be interesting if we are the first to use it. Maybe we could coin the term (like Wearable was coined) :).

tsteur commented:
> What data would be imported? My only frame of reference is web server logs, but this data is still tracked visits, just in a different format. What else would be imported?

We could basically important all analytics related data that is on a third party system and not show any data that was tracked by us. There is currently no item for that yet but one can see it on the roadmap: 

> Data dashboard, Import metrics 

Eg you could define a "Type" named "Shopify", request all sales information etc from there and show only those reports for this type. Another example could be a type "App Store". Here we would fetch/import analytics data from Google Play Store or Apple App Store and show only those reports similar to services like https://appfigures.com . Why would people import such data? To have all data in one place, maybe to connect this data with other "Properties/Resources/...", to reuse features like Row evolution, Alerts, ..., to not having to deal with different UIs etc.

>  could go on, but it's hard to communicate the nuances of a language when so much of it depends on the culture surrounding the language, hopefully my comments give some context.

Yep, that's why I meant I trust you on that :)

diosmosis commented:
> Another example could be a type "App Store". Here we would fetch/import analytics data from Google Play Store or Apple App Store and show only those reports similar to services like https://appfigures.com .

Can you tell me what would be displayed that isn't 'tracked data'? If all the imported data can be described as 'data that was gathered through some form of tracking or collecting', then maybe it makes sense to call the thing that's being tracked a 'Trackable'.

For example, instead of calling a type "Shopify", it could be "Shopify Store" or "Shopify Account". And instead of "App Store" (which technically could be a web app that's being tracked), it could be "Android App", "iPhone App" or even "Mobile App".

tsteur commented:
> Can you tell me what would be displayed that isn't 'tracked data'? If all the imported data can be described as 'data that was gathered through some form of tracking or collecting', then maybe it makes sense to call the thing that's being tracked a 'Trackable'.

Eg when importing data from another shop the data was probably not tracked on that 3rd party system "specifically". The data is just there in the database for different purposes. We just import random data and generate reports from that or we import existing report data from a third party system. I'm not sure how to explain, it's like push (tracking request sent to Piwik) vs pull (data fetched from other system). The platform will be able to do show only reports that were imported by a 3rd party system and it will allow you to show completely different reports compared to what we're used today. It would be rather like `Analyzable` :)

diosmosis commented:
> I'm not sure how to explain, it's like push (tracking request sent to Piwik) vs pull (data fetched from other system).

I don't think the method of data entry is important; just because the data is 'pulled', doesn't mean it wasn't 'pushed' or 'tracked' to begin with.

What I'm having a hard time imagining is exactly what the data is and what would make it different than 'tracked data'. For example, let's say there's a database w/ information about the amount of money in different bank accounts. Some plugin in Piwik is created to display reports on this data, w/o it being 'pushed' by Piwik. The data originally, however, could only be stored by some sort of 'push'. When a new account is created, the amount is stored or when the amount changes, the new amount is stored. Just because Piwik doesn't do the tracking, doesn't mean the data wasn't tracked to begin w/.

What kind of data source isn't build this way? I can't imagine one.

> It would be rather like Analyzable

That's definitely more generic in what it refers to and specific in what it implies will be done to the thing. Probably better for an analytics platform, but I think too many syllables :) `Inspectable` or `Explorable` perhaps... or perhaps not.

Perhaps the issue is that `Trackable` makes it seem like we will always track it and THEN show reports when for some use cases, there is no actual tracking done (at least by Piwik)?

This might also be too generic, but how about `Subject` as in "Subject of Analysis"?

mattab commented:
I'll try to summarise the current discussion:
 * We are looking for a name that makes sense as a class name in PHP, and also this name will be displayed in the UI (eg. left menu), so It should make sense to all users and developers (to us + plugin developers).
 * This name we're looking for is to define a class of objects that will be measured and reported in Piwik. This name will be applied to many "things": Website, Mobile Apps, Apps, Retail store, Smart devices, IoT thing, sensor, a Server, etc.
 * Best possible names so far (in random order)
  * `Trackable`
  * `Resource`
  * `Property` (current one used in PR)
  * `Measurable` (my new suggestion)
 * I'm not sure we need to involve other people at this stage... but if we can't decide on our own, I will be happy to. 
 * Even if we pick the wrong name, it will be easy to change it before 3.0.0 final release. but we are here trying to nail the right name directly so we don't need to come back to it.

What do you think?

tsteur commented:
+1 for `Measurable`. I think that's better than `Resource` or `Trackable`.

> Even if we pick the wrong name, it will be easy to change it before 3.0.0 final release. but we are here trying to nail the right name directly so we don't need to come back to it.

Exactly. By default right now nothing changes. Only if someone enabled the `MobileAppType` plugin the wording changes from `Sites Management` to `XYZ Management`.

diosmosis commented:
Measurable sounds good to me too.

mnapoli commented:
> > Even if we pick the wrong name, it will be easy to change it before 3.0.0 final release. but we are here trying to nail the right name directly so we don't need to come back to it.

> Exactly. By default nothing changes right now. Only if someone enabled the MobileAppType plugin the wording changes from Sites Management to XYZ Management.

Ah well then that solves the problem then :)

I think it's very important for such a large change in the life of Piwik to involve people from marketing/design. If the current implementation isn't visible for users before 3.0 then we can discuss that topic in August. And the class name can be changed later then without caring about BC.

@diosmosis kind of convinced me with the `Trackable` thing (similar to `Wearable`). `Measurable` seems a little too forced but it's more of a feeling.

diosmosis commented:
> `Measurable` seems a little too forced but it's more of a feeling.

I think I understand what you're saying, measuring relates more to individual metrics (ie, measuring length or measuring time) rather than whole reports. Which isn't to say tracking relates to reporting in any way. It just doesn't make you think about measurements vs aggregated reports, so you don't notice it.

I think Measurable still works as viable wording though, and it's definitely better than Property/Resource, IMO ;)

tsteur commented:
So from my understanding we go with `Measurable` and `MeasurableType` for now?

I'd maybe also rename the plugins `WebsiteType` and `MobileAppType` to `WebsiteMeasurable` and `MobileAppMeasurable`? or `WebsiteMeasurableType` and `MobileAppMeasurableType` or keep the names?

mnapoli commented:
> So from my understanding we go with Measurable and MeasurableType for now?

+1

For the plugins we could name them on what they do (what they bring to Piwik): e.g. `WebsiteTracking` and `MobileAppTracking`? `WebsiteMeasurable` & others sound redundant, like `ManHuman` (i.e. a website is a measurable already).

mnapoli commented on core/Property/Property.php:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Provides access to individual properties.
+ */
+class Property extends Site
```
Why does property/measurable extends site (and not the other way around)? Is it only a shortcut for now (will be changed later)?

tsteur replied:
We will most likely change it later. For now not really important. I only want to add this feature on top of the current solution as simple as possible so we can later easily change it again

mnapoli commented on core/Property/PropertySettings.php:
```diff
@@ -0,0 +1,101 @@
+use Piwik\Settings\Setting;
+use Piwik\Type\Type;
+
+class PropertySettings extends Settings
```
`PropertySettings` + `PropertySetting` classes have very confusing names :) (spent a few seconds thinking Github diff had bugged)

tsteur replied:
We will most likely change the naming later once we integrate it properly.

mnapoli commented on core/Site.php:
```diff
@@ -12,6 +12,7 @@
 use Exception;
 use Piwik\Exception\UnexpectedWebsiteFoundException;
 use Piwik\Plugins\SitesManager\API;
+use Piwik\Type\TypeSettings;
```
Un-needed?

mnapoli commented on core/Type/Type.php:
```diff
@@ -0,0 +1,101 @@
+    protected $name = 'General_Property';
+    protected $namePlural = 'General_Properties';
+    protected $description = 'Default type';
+    protected $howToSetupUrl = '';
```
Why provide default properties + default implementation for methods? These will need to be overridden anyway

tsteur replied:
It is currently also the "default type" which would be at some point called "Universal". Once we integrate it properly we have maybe a plugin for that. For simplicity it's here right now.

mnapoli replied:
I see. But having such a thing will probably create some unnecessary code and complexity. It would maybe make sense to me to ignore/not show the `mobileapp` entry if the plugin was removed (which should happen very rarely). (and we could show a notice to the user in that case to explain)

In other words I don't see the point of having a generic "universal" analytics type.

tsteur replied:
See comment here: https://github.com/piwik/piwik/issues/7893#issuecomment-109113915 I think it makes a lot of sense to fall back to a universal type if it is removed and we will still need some kind of wording for it, even if it is ignored etc.

mnapoli commented on core/Type/Type.php:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        return $types;
+    }
```
I'd rather go with a TypeManager service (or whatever the name) rather than static methods so that we can use dependency injection, inject it in other classes and unit test it without having to setup a whole environment (cache, plugin manager, …)

tsteur replied:
We will do that later. Also for reports, widgets etc. For now I'd like to keep things as simple as needed if possible. `Manager` is a wording like `Data` :)

mnapoli replied:
Yes that's why I said whatever the name.

tsteur replied:
I do know that and it's a trade-off. Nothing of this is API and it will be changed soon later again anyway in a most likely very different way. Therefore I'd rather add it in the way we did it everywhere so far. And I would most likely add an integration test for this anyway. Having only unit tests for something is not good either. 

> Same for the plugin manager, we can mock the call and the test will be just as good.

A mock will be never as good as testing the "reality" but it comes close yes and I totally agree that it makes sense sometimes but not always.  

> Yes that means the feature is delivered a little sooner, but that also means that we are introducing code that we will have to refactor 

As mentioned we will very soon remove this code again anyway

>  and there's a high chance the integration test will stay an integration test anyway.

That's nothing bad IMO

Anyway, I can quickly introduce a `Type\Manager` or so. Will change the naming of some classes anyway

mnapoli commented on core/Type/Type.php:
```diff
@@ -0,0 +1,101 @@
+            $types = $cache->fetch($cacheKey);
+        } else {
+            $types = PluginManager::getInstance()->findComponents('Type', '\\Piwik\\Type\\Type');
+            $cache->save($cacheKey, $types);
```
If we cache the PHP instances then we can't inject any dependencies into them? Maybe it's not a big deal to create `WebsiteType` and `MobileAppType` instances on every request (as long as findComponents() is cached)?

tsteur replied:
`getAllTypes` can be called quite often by `getType`. That's why it's currently cached. We could cache in `getType()` but wouldn't make things better. Dependencies will be still injected on creation time if needed and otherwise it shouldn't be needed. At least for now

mnapoli replied:
> getAllTypes can be called quite often by getType

Instances will be created once per request anyway so it doesn't matter (i.e. cache per request). The problem with storing instances in the cache is that there's a high risk of messing up everything because it will also serialize and store dependencies. Storing the types (string) makes more sense.

tsteur replied:
> Instances will be created once per request anyway so it doesn't matter (i.e. cache per request). 

Ok then we can remove the cache

> The problem with storing instances in the cache is that there's a high risk of messing up everything because it will also serialize and store dependencies. Storing the types (string) makes more sense.

It's a transient cache so nothing will be serialized or stored between requests anyway.

tsteur commented:
`WebsiteTracking` and `MobileAppTracking` sounds good

tsteur commented:
Actually, after starting to rename them I'm not sure about `WebsiteTracking` and `MobileAppTracking` anymore. Doesn't it kinda indicate one needs them to track websites or mobile apps? One can still track them without them. Only which reports are presented changes when they are used and some naming. It would be maybe still better than `WebsiteType` but think there would be a better solution.

tsteur commented:
The classes `MeasurableSettings` and `MeasurableSetting` etc. will sound a bit weird as well or not? It indicates we create a setting that is measurable. Was `PropertySettings` and `PropertySetting` before

tsteur committed:
added some ui tests

tsteur commented:
Changed the naming to `Measurables`, added a few UI tests (not too many since we will change it soon again anyway) see https://github.com/piwik/piwik-ui-tests/commit/70c0d772fb74ff3c2c08d1ded9b155517e7b6a2a and applied some feedback from reviews.

Renamed the plugins to `WebsiteMeasurable` and `MobileAppMeasurable` see previous commit. `WebsiteTracking` would be kinda wrong and could lead to false expectations. We can still change it any time. Nothing of it will be public API.

mnapoli commented on core/Measurable/Type/Manager.php:
```diff
@@ -0,0 +1,39 @@
+use Piwik\Plugin\Manager as PluginManager;
+use Piwik\Measurable\Type;
+
+class Manager
```
If we could avoid another `Manager` class that would be great :)

`use Piwik\Plugin\Manager as PluginManager` are so boring to write (e.g. in this very file), and `Manager` type-hints are confusing. `TypeManager` or `MeasurableTypeManager`? (I have no preference between both)

tsteur replied:
I will rename it but we should discuss this in general at some point. I don't mind writing the `as` statements. `MeasurableTypeManager` then we can also just avoid namespaces and put everything in one folder :)

tsteur replied:
Renamed it to `TypeManager`. Btw I often just write `Type\Manager` see eg https://github.com/piwik/piwik/commit/1e9e91ba14b6a5b2aa54b15e31fa382a51cadab3#diff-700dbb21e8c2b2cf9dc0a56a71eb86b4L49

tsteur replied:
Created follow up issue https://github.com/piwik/piwik/issues/8115

mnapoli commented:
one last comment about a class name (above) and looks good to me

mattab commented:
Hi @kaz231  - this is the Pull request that should help you with Mobile App Manager. feel free to ask thomas any question if you have!

tsteur committed:
renamed Manager to TypeManager

tsteur committed:
fix system test
